### PR TITLE
[WIP] Add instrumentation and event recorder

### LINKED ISSUE #127: Events should be recorded into the events API
/kind feature

Events recorded as part of bootstrapping the cluster should be added to the k8s-wide Events API

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:
* Adds instrumentation and tracing across the controllers
  * Uses opencensus.io
* Adds an object event recorder for more structured logging and also updating
  API Server with status.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #127 and #78 

**Special notes for your reviewer**:

Not sure why I ended up doing this, now. Something, something, didn't know what the hell
the machine actuator was upto, something something.

**Definitely not ready**

And not for alpha1v1 in any case.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: cloud/aws/events/events.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+	// Failure event
+	Failure = "Failure"
+
+	// Pending event
+	Pending = "Pending"
+)
+
+var (
+	globalEventRec *record.EventRecorder
```
Comment (chuckha) [inline cloud/aws/events/events.go]: I'm ok if this is a DefaultEventRecorder, similar to how http does DefaultHttpClient, but we should have a way to make a new one and not use mutable globals wherever possible.

What about if the API were like this:

```go
recorder := events.DefaultObjectRecorder
// or
recorder, err := events.NewRecorder(param1, param2)
// check error
objectRecorder, err := events.NewStdObjRecorder(recorder)
// check error
```

File: cloud/aws/instrumentation/instrumentation.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	"strings"
+)
+
+var (
+	AWSOperation, _ = tag.NewKey("aws.operation")
+	AWSService, _   = tag.NewKey("aws.service")
+	AWSRegion, _    = tag.NewKey("aws.region")
+)
+
+func awsRequestAction(r *http.Request) string {
```
Comment (chuckha) [inline cloud/aws/instrumentation/instrumentation.go]: this reads the body, parses the query params, then I'm pretty confused on line 50, why does this put the body back into a reader that isn't closable? it then returns the query value for the Action paramter

Comment (randomvariable) [reply cloud/aws/instrumentation/instrumentation.go]: Just experimenting at the moment. Was hacking around to get the body without destroying the request. It doesn't yet actually do the thing I want it to.

Comment (chuckha) [reply cloud/aws/instrumentation/instrumentation.go]: ohh ok, didn't realize this was experimental code. hack on!

**Comment by chuckha** [comment]
ah ha, totally missed https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/78 even though it was linked in the description

File: cloud/aws/events/events.go (hunk 1 above)
Comment (randomvariable) [reply cloud/aws/events/events.go]: Hah, yeah. Felt dirty writing that.

**Comment by randomvariable** [comment]
Sorry, deliberately didn't say much on this PR so as to not distract from the MVP stuff.

Basically, I was wondering how easy it would be to monitor the AWS calls so that we could easily spot things like duplicate RunInstances when I got lost in the weeds of the machine actuator.

> Could you break this into several PRs?
* Yes. More of a POC than anything.

Think the order of importance here is:
1. Event recorder, as that'll populate the objects when people do a `kubectl describe`.
2. Instrumentation with either Prometheus client library or OpenCensus
3. Tracing with either OpenCensus, OpenTracing or other.

On the OpenCensus vs. OpenTracing stuff - don't think they're exactly the same thing. OpenTracing is pretty much client libraries + a specific protocol, whereas OpenCensus is a client library with pluggable protocols, one of which happens to be Prometheus.

The instrumentation's more important than tracing, and so the OpenCensus vs. Prometheus client library is probably more salient. Think I'll try both and see what they're like.

Advantage of the former is support for exporting metrics to CloudWatch & also getting support for tracing in a single library for those who want an all-on-AWS experience.

In terms of API stability, neither Prometheus client or the OpenCensus library for Go has reached version 1, but the latter is more in flux (hence pinning an actual commit is a constraint in Gopkg.toml.

Also, from a previous Twitter conversation, @rakyll was happy to assist.

**Comment by randomvariable** [comment]
To return in a later PR...

**Pull request closed without merging**